### ISSUE TITLE
Binary Ninja 3.3 has broken antialiasing with Windows 10 (no HiDPI)

### ISSUE THREAD

**Issue by amyspark** [Vector35/binaryninja-api#3817]
**Version and Platform (required):**
- Binary Ninja Version: 3.3.3996 Personal (Build ID e34a955e)
- OS: Windows 10
- OS Version: 21H2
- CPU Architecture x64

**Bug Description:**

The font antialiasing has been failing for me since at least 3.2. While I could work around it in that version (see below), 3.3 has broken it completely. having blurry widgets even with the default Inter typeface.

**Steps To Reproduce:**
Please provide all steps required to reproduce the behavior:
1. Go to Settings, type `font` in the search bar.
2. Set font.app > "Application font name" to "Segoe UI"
3. Restart Binary Ninja.
4. Repeat 1, then set font> "Font Antialiasing Style" to any desired option.

**Expected Behavior:**
Crisp font rendering.

**Screenshots:**

With 3.3:

|              | Inter/Roboto | Segoe UI/Consolas |
| --------- | ------------- | ------------- |
| subpixel |  ![imagen](https://user-images.githubusercontent.com/13498015/213881212-dc58f161-f029-4dc0-ae45-6c9712c7bdb5.png)  | ![imagen](https://user-images.githubusercontent.com/13498015/213881264-b424f8ac-79ae-4d03-8230-5573924ec08b.png)  |
| grayscale | ![imagen](https://user-images.githubusercontent.com/13498015/213881137-65e3e6a5-c60a-4b3b-a8de-54b531bccc71.png)  | ![imagen](https://user-images.githubusercontent.com/13498015/213881286-53bd6bf2-01aa-4f11-8325-555ee320e30e.png)  |
| none      |   ![imagen](https://user-images.githubusercontent.com/13498015/213881173-09198588-4d9d-4f51-908d-7d475e90460d.png)   | ![imagen](https://user-images.githubusercontent.com/13498015/213881306-0c7139f7-07d7-4489-ae4a-7cf15f5b78b0.png)   |

With 3.2:

|              | Inter/Roboto | Segoe UI/Consolas |
| --------- | ------------- | ------------- |
| subpixel |  ![imagen](https://user-images.githubusercontent.com/13498015/213881738-0f8b054d-d5ea-4d66-be1e-11bc29a41a6e.png)  | ![imagen](https://user-images.githubusercontent.com/13498015/213881680-6f4feb0b-9002-4c1b-907d-e042dd20fd1d.png)  |
| grayscale | ![imagen](https://user-images.githubusercontent.com/13498015/213881760-688bdb0d-1049-48c0-89f8-39d14c0e33a7.png)  | ![imagen](https://user-images.githubusercontent.com/13498015/213881649-c8273955-89d4-4a8c-8558-67d3b35cfadc.png)  |
| none      |  ![imagen](https://user-images.githubusercontent.com/13498015/213881786-998d00b3-f776-483e-a0ae-cfaf1b03b30d.png)  | ![imagen](https://user-images.githubusercontent.com/13498015/213881622-4d077445-36e7-42d1-acca-47cacdf04b87.png)   |

**Additional Information:**

In 3.2, it was possible to work around this by setting "Segoe UI" and "Consolas" as the application and view font name respectively, plus setting the antialiasing to Subpixel.
In 3.3, this has broken completely and the UI look just wrong irrespective of font and antialiasing option, as shown in the screenshots above.

**Comment by fuzyll** [comment]
Could you give us some more information about your specific setup? I've tried replicating this on both Windows 10 and Windows 11 VMs, plus a bare-metal Windows 11 install, and I don't think I'm seeing the same font issues that you are.

I tossed Segoe UI in as my font choice and then took two screenshots from inside of my Windows 10 VM:

![Capture](https://user-images.githubusercontent.com/607452/214471833-6053abf0-b418-4a76-afe2-b0fe4fcaf5bf.PNG)

![Capture2](https://user-images.githubusercontent.com/607452/214471842-d437793f-9c69-4c44-b92d-bcb2b0e481c2.PNG)

This is with subpixel antialiasing on.

Now, I'm usually pretty nitpicky about fonts and rendering and stuff, but these seem reasonable to me? They also seem slightly clearer than your own screenshots? But, I dunno, maybe I'm missing something.

**Comment by fuzyll** [comment]
For completeness, here's how my bare-metal Windows 11 install looks with 3.2:

<img width="700" alt="Screenshot 2023-01-24 223342" src="https://user-images.githubusercontent.com/607452/214474273-09ea593f-ecbf-4668-8516-459e359381c7.png">

...and 3.3:

<img width="775" alt="Screenshot 2023-01-24 224146" src="https://user-images.githubusercontent.com/607452/214475067-33db797a-d725-4f02-b4c0-de371a9d84ed.png">

...both with Segoe UI and subpixel antialiasing enabled.

**Comment by amyspark** [comment]
> For completeness, here's how my bare-metal Windows 11 install looks with 3.2:
> 
> <img alt="Screenshot 2023-01-24 223342" style="image-rendering: crisp-edges;" src="https://user-images.githubusercontent.com/607452/214474273-09ea593f-ecbf-4668-8516-459e359381c7.png">
> 

That one already confirms my issue, your 11 VM has no antialiasing:

![imagen](https://user-images.githubusercontent.com/13498015/214568670-32154277-c242-4f33-8f7b-275638b3b5ac.png)

whereas 3.2 shows the smoothed edges of subpixel Cleartype:

![imagen](https://user-images.githubusercontent.com/13498015/214568799-1613ee63-8c99-458b-ac0e-c6f5a43a2b3a.png)

Your 10 VM on the other hand looks OK:

![imagen](https://user-images.githubusercontent.com/13498015/214570327-895fec75-6727-43eb-9f2d-b16a76995b78.png)

My setup is as follows:

- OS: Windows 10 21H2
- CPU: Ryzen 7 2700x
- GPU: Nvidia Geforce GTX 1660 Ti, driver version 527.56
- Screens: 1080p, 1x Samsung S24F350, 1x LG M2380A, no HiDPI. Both calibrated with ArgyllCMS and with the ICC profile active

I'll check GPU driver settings and ICC profile now and let you know.

**Comment by fuzyll** [comment]
Thanks, I now understand better what you're talking about. And, actually, I think you might've already helped me understand the problem better. The screenshots I had from Windows 11 are actually _not_ from a virtual machine, which suggests the reason why we didn't catch this before release is that there is a difference between the rendering inside the VM versus bare-metal.

I also have an Nvidia Geforce GTX card in the bare-metal machine I got the Windows 11 screenshots from. I'm wondering if there's an interaction with Qt and device drivers that is causing the issue, since we haven't changed any font rendering code on our end.

**Comment by amyspark** [comment]
>  a difference between the rendering inside the VM versus bare-metal

I've not digged into Binja's innards, but are you shipping Qt6? Perhaps it's high DPI antialiasing rearing its head, I believe it's enabled by default there.

**Comment by fuzyll** [comment]
We are, actually. A quick 1-minute Google search didn't bring up anything in the way of a recent Qt bug regarding high DPI antialiasing. Are you aware of something I could try to resolve the issue?

We do have the following `ifdef` that _shouldn't_ be compiling into the product now:

```
#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
	QCoreApplication::setAttribute(Qt::AA_EnableHighDpiScaling);
	QCoreApplication::setAttribute(Qt::AA_UseHighDpiPixmaps);
#endif
```

...presumably because it _is_ on by default now. I'm not personally very knowledgeable about antialiasing in Qt on Windows, though.

If you're aware of some stuff I could try, please let me know, otherwise I'll try and mess with some settings when I get a chance and see what happens.

**Comment by fuzyll** [comment]
Does this look better?

![image](https://user-images.githubusercontent.com/607452/218024179-80c784c2-82c1-4762-9c45-a3cb1f4be794.png)

**Comment by amyspark** [comment]
> Does this look better?
> 
> 
> I think it does, but I'm obviously not as sensitive, so I want to make sure before I try and figure out how to get this out in a `dev` build.

I think so, yes! I need to test with the system fonts, but the borders look as they should. Let me know if/when I can check it out.

**Comment by fuzyll** [comment]
This is now live in builds >= 4060 on the `dev` branch and will be included in the next stable release.

For posterity, there is some problem with Qt 6's default High DPI rendering on Windows. The solution was to put this:

```c++
qputenv("QT_ENABLE_HIGHDPI_SCALING", "0");
```

...before the main `QApplication` gets instantiated.

Big thanks to Stuart Mentzer's post [here](https://sourceforge.net/p/qwt/bugs/339/) that clued me in to this as a possible solution (and pointed out that the solution I tried first probably wasn't going to work). Additional shout out to the [Qt 6 documentation](https://doc.qt.io/qt-6/highdpi.html) that does, in fact, explain how this works...but does so buried at the bottom of the page.

**Comment by fuzyll** [comment]
I am reverting the change I made. I'm reasonably certain it's behind a different bug where the Windows UI appears to be super small relative to the text on the screen. If I'm correct, I'll need to find a different way to make this happen.

**Comment by amyspark** [comment]
I think I have an idea for you @fuzyll, can you check which backend you're using for rendering binja's UI? In particular, are you using FreeType?

**Comment by amyspark** [comment]
Also see https://bugreports.qt.io/browse/QTBUG-100774 and related (coming from https://github.com/telegramdesktop/tdesktop/issues/25936 where I'm experiencing a related kind of blurry once FreeType is enabled).

**Comment by fuzyll** [comment]
Thanks! Ironically, I had just found that same issue (and the one it links to, [111380](https://bugreports.qt.io/browse/QTBUG-111380) last night, along with [this one](https://bugreports.qt.io/browse/QTBUG-103488), which all seem related. This is unfortunately not my top priority right now, but I'm still slowly getting to understand the High DPI code and trying to pick through where the problem is. Unfortunately, since they've removed the other settings and replaced them with a single environment variable switch, figuring this out is slightly painful.

Appreciate you helping to point me in the right direction, though! Hopefully I'll have time to sit down and figure it out for good soon. 🙂

**Comment by fuzyll** [comment]
I'm slipping this to next release. I suspect there are multiple confounding factors here, only one of which is the antialiasing being off.

We upgraded to Qt 6.4.3 for the next release, which has had _some_ improvements, but linear view in particular still looks pretty bad:

![image](https://user-images.githubusercontent.com/607452/234378279-9b37e300-7c28-4604-80b6-dcc71b71ca41.png)

There are, however, places in the product that don't look quite so bad. So, I might need to do a more thorough audit of how we're rendering text to see if maybe we're causing additional issues. Since some of the Qt issues mentioned earlier are still open as well, it's also possible that a future update to the Qt 6.5 series (or newer) might also help resolve some of these problems.

I apologize for this process taking so long, and I do want to get it fixed. It's just been difficult because of so many confounding factors.

**Comment by amyspark** [comment]
@fuzyll I've been digging into old bug reports, and I found this: https://bugs.chromium.org/p/chromium/issues/detail?id=405445
Together with https://stackoverflow.com/a/75839785, I'm wondering if Qt6 needs to have a whitelist of DW-unhappy fonts, or alternatively some custom hinting parameters?

**Comment by fuzyll** [comment]
Hey thanks!

So, I'm not 100% sure if the top one is relevant or not... The link seems to suggest that if we were having an issue where DirectWrite was unable to display the font correctly, we should be getting some sort of 'halo' effect because of how the colors bits are swapped around? But, regardless, DirectWrite is something I need to look into because Qt6 now relies heavily (entirely?) on that for rendering on Windows.

The second link was interesting and had me looking all over the codebase for places where we mess with the font hinting settings. It turns out there were some inconsistencies, so hopefully I can make an improvement here regardless. But, I'm not _completely_ sold that the setting it encourages turning on is the right move. Here's how it looks in a special build where we go all-in on platform font hinting:

With `subpixel`:

![image](https://github.com/Vector35/binaryninja-api/assets/607452/0491e889-2aec-468e-ae9a-575191bc8bb4)

With `hidpi`:

![image](https://github.com/Vector35/binaryninja-api/assets/607452/af646575-2f5a-41e8-bf88-691f62e7bafd)

With `none`:

![image](https://github.com/Vector35/binaryninja-api/assets/607452/df831439-5da6-4a78-bd76-9afad0df56ba)

None of these look 'blurry'; which is great! But, here's how Firefox is rendering your comment on the same box:

![image](https://github.com/Vector35/binaryninja-api/assets/607452/64d218b4-4c55-4dd7-8505-3c613c7bd6cb)

...and I just can't shake the feeling that we're still getting something very wrong on Windows on HiDPI screens. I haven't tried this in a VM, but based on earlier screenshots in this ticket, I think this should look even better in a VM. It's bare-metal font rendering that's the problem.

Anyway, thanks for the links! I'm still looking into this in-between other things, and I appreciate any other information you're able to give me. Hopefully we'll figure this out and get it fixed soon!

**Comment by amyspark** [comment]
YESSS! 🎉 That looks awesome @fuzyll !

Let me know once it lands in the development channel and I'll be more than happy to test it.

**Comment by fuzyll** [comment]
Okay, these changes are now live on builds >= 3.5.4425-dev.

There is still an issue with the default interface font (Inter), but only in certain situations, and I'm starting to think that maybe there's an issue with that font interacting with the `PreferFullHinting` setting. If hinting is turned _off_, it looks great (which is how I was able to fix the way the font looks in the interface in general), but if it's turned _on_ (which happens for any widgets that get dynamically created) it looks really strange.

The problem with the hinting setting is that it is _global_ - once you set it, it's that way for everything rendered from then on. So, the current workaround is to turn it off, get a bunch of the interface constructed, and then turn it on so that Linear View looks good. Unfortunately, this means stuff like the Plugin Manager can look a bit off for much of the rendered text.

I'll keep digging into the issue, but for now this should be fixed _enough_ that Binary Ninja is pleasant to work in on Windows now.

**Comment by amyspark** [comment]
@fuzyll Looks great now! One nit: Segoe UI at sizes between 8 and 11 points look *very* blurry still, I think DirectWrite or something along the pipeline is doing upscaling of a lower integral size. That's the only thing outstanding now.

**Comment by fuzyll** [comment]
Just to confirm: When you say it looks "very blurry", where are you looking? I'm actually not seeing any issues with Segoe UI at 10 point on 3.5.4426-dev, which maybe suggests that the problem I'm still trying to track down is actually the fault of our Inter font file and not something in Qt.

I'm currently at 150% scale:

![image](https://github.com/Vector35/binaryninja-api/assets/607452/89170581-6a91-426f-9021-43ba60626ed7)

...and this is how Segoe UI looks on the New Tab page:

![image](https://github.com/Vector35/binaryninja-api/assets/607452/dd7c9220-fa91-45ab-a84a-21468227e971)

It also looks fine in the Plugin Manager:

![image](https://github.com/Vector35/binaryninja-api/assets/607452/d017c551-bff7-48ff-b868-7b960a48822d)

...which *does not* look fine in our Inter font right now (because hinting appears to be off, despite it actually being turned on):

![image](https://github.com/Vector35/binaryninja-api/assets/607452/86ed4c77-e53d-4e20-b67a-6af635ce2b2c)

**Comment by amyspark** [comment]
This is with Segoe UI at the default size (10 point):

![imagen](https://github.com/Vector35/binaryninja-api/assets/13498015/1213e938-65e8-4edb-a7c2-9ba852534bc2)

I also took a pic with my camera showing the blurry font vs Consolas:

![imagen](https://github.com/Vector35/binaryninja-api/assets/13498015/ee9353ed-54db-4493-953a-5bd5fb02afcd)

**Comment by fuzyll** [comment]
Okay, I've made some more changes and I believe this may actually be taken care of now? Hopefully? Maybe?

If you could check out version 3.5.4481-dev (which should be making its way to update servers as I write this), that would be appreciated.

The TL;DR is that I realized most of what we were doing was working around a prior Qt 5 bug and an actual issue with the custom-built version of Inter that we were shipping with Binary Ninja. As a result, I've ripped out nearly all of the custom code we had surrounding Windows font handling and put us back to Qt defaults. I've also shipped a new version of the font that was built with hints for DirectWrite.

The combination of all these changes _should_ have all font issues resolved on Windows, regardless of DPI. I've also changed the handling of one of the font antialiasing options to prefer antialiasing (where previously it was left to the Qt default of "no preference"). This can still be turned down/off in our settings, so if things still seem a bit blurry by default, give the other options a shot again and see if that resolves things.

But, from my testing across all 3 platforms and in VMs vs. not in VMs, this should result in _either_ no change (macOS), a slight improvement (Linux, Windows on low DPI/in VMs), or a huge improvement (Windows at high DPI) in font rendering across the product.

Absolutely let me know if this isn't the case because there are a fair number of changes and my testing, while fairly extensive, wasn't exhaustive, so there's always a chance I missed something. Hoping I can finally close this issue! 🤞

**Comment by amyspark** [comment]
Hey! 🎉  yes, it works!

Nit: did you hardcode TTF Inter as the UI font? It correctly recognises the Segoe UI sizes now, but the main view uses Inter nomatter what I set.
![imagen](https://github.com/Vector35/binaryninja-api/assets/13498015/69996f9d-3e38-4bf8-9c84-7b1e82370203)

**Comment by fuzyll** [comment]
I did not, but I did implement a fallback to our font if there isn't a match on the filesystem for the font specified in the user settings to fix #4566. So, it sounds like I did indeed have an edge-case that I didn't account for and I'll need to implement that differently. 😞

I'll close this one and re-open that one while I look for a better solution. So close! 😛 Thanks again for testing this out for me.
